core/async/src/test_loop/delay_sender.rs

```diff
@@ -0,0 +1,69 @@
+use std::{sync::Arc, time::Duration};
+
+use crate::messaging;
+
+use super::multi_instance::IndexedDelaySender;
+
+/// Interface to send an event with a delay (in virtual time). It can be
+/// converted to a Sender for any message type that can be converted into
+/// the event type, so that a DelaySender given by the test loop may be passed
+/// to production code that expects a Sender.
+pub struct DelaySender<Event> {
+    // We use an impl object here because it makes the interop with other
+    // traits much easier.
+    im: Arc<dyn CanSendWithDelay<Event>>,
+}
+
+impl<Message, Event: From<Message> + 'static> messaging::CanSend<Message> for DelaySender<Event> {
+    fn send(&self, message: Message) {
+        self.send_with_delay(message.into(), Duration::ZERO);
+    }
+}
+
+/// Underlying trait for DelaySender.
+pub(crate) trait CanSendWithDelay<Event>: Send + Sync {
+    fn send_with_delay(&self, event: Event, delay: Duration);
+}
+
+/// Implements any CanSendWithDelay<Other> as long as Other can be converted into Event.
+pub struct NarrowingDelaySender<Event> {
```

Comment:
should this really be pub?

Reply:
You're right. (I got rid of this struct though).